### ISSUE TITLE
NRM2 Runtime Inconsistency

### ISSUE THREAD

**Issue by dasswap** [peterahrens/ReproBLAS#1]
Just wanted to run through some inquiries about ReproBLAS (in particular about the NRM2 operations).

From our analysis of the RDNRM2 and RDZNRM2, we noticed that the time taken for complex operations quadrupled the time taken for real operations. We can't really understand why that would be the case - we expected it to be a doubling (since we would do twice as many operations for complex compared to real operations). Could you explain why that is the case given the current code?

We looked at the code (in particular, dmdssq.coog and dmzssq.coog in idxdBLAS) and it did look like it was twice as many operations for complex. It occurred to us that we could simply square the real and complex parts of the numbers and sum them before updating the bins -  this could save up to half the time taken. What do you think are the pros and cons of this approach? 

For reference, below is the Excel file of the time analysis undertaken - refer to sheet NRM2 Comparison in particular to observe the weird phenomenon in question.

[ReproBLAS Testing.xlsx](https://github.com/peterahrens/ReproBLAS/files/1192119/ReproBLAS.Testing.xlsx)

**Comment by willow-ahrens** [comment]
So it seems there are two questions here.
1. Why is there weird performance?
2. Why not sum the real and complex contributions of the square before reproducible accumulation?

To answer 1:

   I suggest you first get rid of the crazy autotuned code and use the simpler default arguments. To change the content of the .ccog files to use the simpler default_args.json, use the command `make update ARGS=src/default_args.json` This will make the code easier to read while still being correct (the code won't be unrolled like 37 times as it is now).

   Next, see where the time is being spent. Profile the code with something that can give you line-by-line time usage. If you can get it to run on Mac, I suggest you use instruments (you'll need compiler flags -pg and perhaps something else to get it to link in source files. Let me know if you get stuck with this). If you want more of a challenge/learning experience, you can try intel vtune or something of the sort on BlueWaters.

To answer 2:

We don't sum the contributions as floating point numbers before adding them to our accumulator because this could result in a loss of accuracy. Consider the norm of the vector

<1 + i2<sup>-35</sup>, 0 + i>

Squaring the entries of the vector yields

<1 - 2<sup>-70</sup>, 0 - 1>

If we add the real and imaginary contributions together with floating point `double` precision first, we get the vector

<1, -1>

since `double` has only 53 bits of significand, and we get a final answer of 0 when we sum the entries with reproducible summation.

However, if we add each contribution separately with reproducible summation, we get a final answer of -2<sup>-70</sup> since ReproBLAS is awesome and can be thought of as having 80 bits of precision (its a little more nuanced than this, but this is essentially the case).

**Comment by willow-ahrens** [comment]
As far as I can tell, this issue is closed by fbb1cfb and 663e1d4. 

The tuned_args.json file generated by the autotuner had a bizarre block size of 64 (it should have been around 1024) and suboptimal max_unroll_width and max_pipe_width parameters. Setting these parameters by hand resulted in performance almost identical to the double precision routine.

I am surprised that the autotuner was so wrong in this case. This may be indicative of some possible problems with the autotuner (or perhaps it was just a fluke in the search algorithm for just this one routine).

You can get the fixed version of ReproBLAS on the master branch.

```
Double Precision Time(s):

+------+------+------------+------------+
|  N   | fold |   RDNRM2   |   DAMAX    |
+======+======+============+============+
| 1024 | 3    | 1.1303e-06 | 2.4290e-07 |
+------+------+------------+------------+
| 4096 | 3    | 3.8824e-06 | 1.0711e-06 |
+------+------+------------+------------+

Complex Double Precision Before tuned_args.json Update Time(s):

+------+------+------------+------------+
|  N   | fold |  RDZNRM2   |   ZAMAX    |
+======+======+============+============+
| 512  | 3    | 2.0027e-06 | 2.3506e-07 |
+------+------+------------+------------+
| 2048 | 3    | 6.9276e-06 | 1.0122e-06 |
+------+------+------------+------------+

Complex Double Precision After tuned_args.json Update Time(s):

+------+------+------------+------------+
|  N   | fold |  RDZNRM2   |   ZAMAX    |
+======+======+============+============+
| 512  | 3    | 1.4184e-06 | 2.7846e-07 |
+------+------+------------+------------+
| 2048 | 3    | 4.0150e-06 | 1.0211e-06 |
+------+------+------------+------------+
```
